cmd/cri-dockerd/app/server.go

```diff
@@ -0,0 +1,146 @@
+/*
+Copyright 2015 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package app
+
+import (
+	"fmt"
+	"net/url"
+
+	"k8s.io/component-base/cli/flag"
+	"k8s.io/klog"
+	"k8s.io/kubernetes/cmd/cri-dockerd/app/options"
+	"k8s.io/kubernetes/pkg/kubelet/dockershim"
+	dockerremote "k8s.io/kubernetes/pkg/kubelet/dockershim/remote"
+	"k8s.io/kubernetes/pkg/kubelet/server/streaming"
+	utilflag "k8s.io/kubernetes/pkg/util/flag"
+	"k8s.io/kubernetes/pkg/version/verflag"
+
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+)
+
+const (
+	componentDockerCRI = "DockerCRI"
+)
+
+// NewDockerCRICommand creates a *cobra.Command object with default parameters
+func NewDockerCRICommand(stopCh <-chan struct{}) *cobra.Command {
+	cleanFlagSet := pflag.NewFlagSet(componentDockerCRI, pflag.ContinueOnError)
+	cleanFlagSet.SetNormalizeFunc(flag.WordSepNormalizeFunc)
+	kubeletFlags := options.NewDockerCRIFlags()
+
+	cmd := &cobra.Command{
+		Use:  componentDockerCRI,
+		Long: `CRI that connects to Docker Daemon`,
+		// The cri-dockerd has special flag parsing requirements to enforce flag precedence rules,
+		// so we do all our parsing manually in Run, below.
+		// DisableFlagParsing=true provides the full set of flags passed to the cri-dockerd in the
+		// `args` arg to Run, without Cobra's interference.
+		DisableFlagParsing: true,
```

Comment:
Is this still required? Does `cri-dockerd` need `KubeletConfiguration` to operate or it needs only the docker shim related command line flags of the kubelet?

Reply:
@rosti until we move `cri-dockerd` into its own repo, it seemed best to keep this stuff the same between "regular" kubelet and "cri-dockerd" in terms of command line flags and config parsing.

Reply:
we need to figure out how to tell the users to install the new `cri-dockerd` binary.

Reply:
@dims Indeed, I can see that this outside the scope of this PR, but we can still add a simplification TODO there.
The complex way of doing command line parsing in the kubelet is required, because of the ability to override config fields with command line flags. If we don't need the KubeletConfiguration in cri-dockerd, then we can greatly simplify things.

@neolit123 that can be done by adding a dependency in the kubelet package to the cri-dockerd package.

Reply:
i think the kubelet with dockershim and cri-dockerd have to be supported in parallel for a while.

first cri-dockerd has to become a part of the node packages and the kubernetes release process and only once the first classed deployers like kops, kubeadm etc switch to it we can remove dockershim.

i have concerns making the hard switch within one release as it may bust existing setups.